docs/resources/server/enchantments/index.md

```diff
@@ -0,0 +1,307 @@
+  // or a list of enchantment ids, such as ["minecraft:sharpness", "minecraft:fire_aspect"],
+  // or enchantment tag, such as "#examplemod:exclusive_to_enchant_name".
+  // Incompatible enchantments will not be added to the same item by vanilla mechanics.
+  "exclusive_set": "#examplemod:exclusive_to_enchant_name",
+  
+  // The likelihood that this enchantment will appear in the Enchanting Table. 
+  // Bounded by [1, 1024].
+  "weight": 6,
+  
+  // The maximum level this enchantment is allowed to reach.
+  // Bounded by [1, 255].
+  "max_level": 3,
+  
+  // The maximum cost of this enchantment, measured in "enchanting power". 
+  // This corresponds to, but is not equivalent to, the threshold in levels the player needs to meet to bestow this enchantment.
+  // See below for details.
+  // The actual cost will be between this and the min_cost.
+  "max_cost": {
+    "base": 45,
+    "per_level_above_first": 9
+  },
+  
+  // Specifies the minimum cost of this enchantment; otherwise as above.
+  "min_cost": {
+    "base": 2,
+    "per_level_above_first": 8
+  },
+
+  // The cost that this enchantment adds to repairing an item in an anvil in levels. The cost is multiplied by enchantment level.
+  // If an item has a DataComponentTypes.STORED_ENCHANTMENTS component, the cost is halved. In vanilla, this only applies to enchanted books.
+  // Bounded by [1, inf).
+  "anvil_cost": 2,
+  
+  // (Optional) A list of slot groups this enchantment provides effects in. 
+  // A slot group is defined as one of the possible values of the EquipmentSlotGroup enum.
+  // In vanilla, these are: `any`, `hand`, `mainhand`, `offhand`, `armor`, `feet`, `legs`, `chest`, `head`, and  `body`.
+  "slots": [
+    "mainhand"
+  ],
+
+  // The effects that this enchantment provides as a map of enchantment effect components (read on).
+  "effects": {
+    "examplemod:custom_effect": [
+      {
+        "effect": {
+          "type": "minecraft:add",
+          "value": {
+            "type": "minecraft:linear",
+            "base": 1,
+            "per_level_above_first": 1
+          }
+        }
+      }
+    ]
+  }
+}
+```
+
+### Enchantment Costs
+
+The `max_cost` and `min_cost` fields specify boundaries for how much enchanting power is needed to create this enchantment. There is a somewhat convoluted procedure to actually make use of these values, however.
+
+First, the table takes into account the return value of `IBlockStateExtension#getEnchantPowerBonus()` for the surrounding blocks. From this, it derives a 'base cost' for each slot. This cost is shown in-game as the green numbers besides the enchantments in the menu. For each enchantment, the base cost is modified twice by a random value derived from the item's enchantability (its return value from `IItemStackExtension#getEnchantmentValue()`), like so:
+
+`(Modified Cost) = (Base Cost) + random.nextInt(e / 4 + 1) + random.nextInt(e / 4 + 1)`, where `e` is the enchantability score.
```

Comment:
I think they way you are wording this is wrong. The costs are calculated from `EnchantmentHelper#getEnchantmentCost`. Using those costs, a random enchantment is selected from an available enchantment list that can be obtained for that cost via `EnchantmentHelper#selectEnchantment`.

What you're describing here is how the enchantment level is chosen, not the cost.

Reply:
ParchmentMC calls the array of these values 'costs', which is where I got the idea to call them that, but I can see that it makes more sense to call these enchantment levels.